File: 백준/Gold/14719. 빗물/빗물.py
def solution(w: int, blocks: list):
    answer = 0
    highest = max(blocks)

    # 정방향
    now_highest = 0
    stop_index1 = 0
    for i in range(w):
        if blocks[i] >= highest:
            stop_index1 = i
            break
        elif blocks[i] < now_highest:
            answer += now_highest - blocks[i]
        else: # blocks[i] > now_highest:
            now_highest = blocks[i]

    # 역방향
    now_highest = 0
    stop_index2 = 0
    for i in range(w - 1, -1, -1):
        if blocks[i] >= highest:
            stop_index2 = i
            break
        elif blocks[i] < now_highest:
            answer += now_highest - blocks[i]
        else:
            now_highest = blocks[i]

    # 중간 부분 계산
    for i in range(stop_index1 + 1, stop_index2):
        answer += highest - blocks[i]

    print(answer)

h, w = map(int, input().split())
blocks = list(map(int, input().split()))
solution(w, blocks)

File: 백준/Gold/14719. 빗물/빗물.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        br.readLine();
        StringTokenizer st = new StringTokenizer(br.readLine());
        int[] blocks = new int[st.countTokens()];
        for (int i = 0; i < blocks.length; i++) {
            blocks[i] = Integer.parseInt(st.nextToken());
        }

        System.out.println(solution(blocks));
    }

    static int solution(int[] blocks) {
        int answer = 0;

        for (int i = 1; i < blocks.length - 1; i++) {
            int leftHighest = getMax(blocks, 0, i);
            int rightHighest = getMax(blocks, i + 1, blocks.length);

            int compare = Math.min(leftHighest, rightHighest);

            if (compare > blocks[i]) {
                answer += compare - blocks[i];
            }
        }

        return answer;
    }

    private static int getMax(int[] arr, int start, int end) {
        int max = Integer.MIN_VALUE;
        for (int i = start; i < end; i++) {
            if (arr[i] > max) {
                max = arr[i];
            }
        }
        return max;
    }
}